search/collectors/collector_heap.go

```diff
@@ -0,0 +1,241 @@
+	"golang.org/x/net/context"
+)
+
+type HeapCollector struct {
+	size          int
+	skip          int
+	total         uint64
+	maxScore      float64
+	took          time.Duration
+	sort          search.SortOrder
+	results       search.DocumentMatchCollection
+	facetsBuilder *search.FacetsBuilder
+
+	lowestMatchOutsideResults *search.DocumentMatch
+}
+
+var COLLECT_CHECK_DONE_EVERY = uint64(1024)
+
+func NewHeapCollector(size int, skip int, sort search.SortOrder) *HeapCollector {
+	hc := &HeapCollector{size: size, skip: skip, sort: sort}
+	heap.Init(hc)
+	return hc
+}
+
+func (hc *HeapCollector) Collect(ctx context.Context, searcher search.Searcher, reader index.IndexReader) error {
+	startTime := time.Now()
+	var err error
+	var next *search.DocumentMatch
+
+	// search context with enough pre-allocated document matches
+	// we keep references to size+skip ourselves
+	// plus possibly one extra for the highestMatchOutsideResults
+	// plus the amount required by the searcher tree
+	searchContext := &search.SearchContext{
+		DocumentMatchPool: search.NewDocumentMatchPool(hc.size + hc.skip + 1 + searcher.DocumentMatchPoolSize()),
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		next, err = searcher.Next(searchContext)
+	}
+	for err == nil && next != nil {
+		if hc.total%COLLECT_CHECK_DONE_EVERY == 0 {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+			}
+		}
+		if hc.facetsBuilder != nil {
+			err = hc.facetsBuilder.Update(next)
+			if err != nil {
+				break
+			}
+		}
+
+		err = hc.collectSingle(searchContext, reader, next)
+		if err != nil {
+			break
+		}
+
+		next, err = searcher.Next(searchContext)
+	}
+	// compute search duration
+	hc.took = time.Since(startTime)
+	if err != nil {
+		return err
+	}
+	// finalize actual results
+	err = hc.finalizeResults(reader)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (hc *HeapCollector) collectSingle(ctx *search.SearchContext, reader index.IndexReader, d *search.DocumentMatch) error {
+	// increment total hits
+	hc.total++
+	d.HitNumber = hc.total
+
+	// update max score
+	if d.Score > hc.maxScore {
+		hc.maxScore = d.Score
+	}
+
+	var err error
+	// see if we need to load ID (at this early stage, for example to sort on it)
+	if hc.sort.RequiresDocID() {
+		d.ID, err = reader.FinalizeDocID(d.IndexInternalID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// see if we need to load the stored fields
+	if len(hc.sort.RequiredStoredFields()) > 0 {
+		if d.ID == "" {
+			// look up the id since we need it for lookup
+			d.ID, err = reader.FinalizeDocID(d.IndexInternalID)
+			if err != nil {
+				return err
+			}
+		}
+		d.Document, err = reader.Document(d.ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	// optimization, we track lowest sorting hit already removed from heap
+	// with this one comparision, we can avoid all heap operations if
+	// this hit would have been added and then immediately removed
+	if hc.lowestMatchOutsideResults != nil {
```

Comment:
I removed the lower bound optimisation that was here before...

Great that it's back!